### PR TITLE
Add <p align> support to HtmlText

### PR DESCRIPTION
**Pull request by Yanrishatum**
What the title says. 
Additionally, XML now parsed only once instead of twice.
Lines are now calculated always, regardless of default Align.
Update: Reworks how Text initialize glyphs in general.

### CODE REVIEW

**Comment by Yanrishatum** [comment]
Marking as WIP, because it seems that there some issues still present in some cases I encounted today.

**Comment by Klug76** [comment]
Looks like closed tag does not restore previous alignment? `<p align='left'>left<p align='center'>center</p></p>left`

**Comment by Yanrishatum** [comment]
Oh boy, this took a while. I've basically reworked how `initGlyphs` operate a lot.
Currently (i.e. in `master`) Text will iterate over entire text 4 times, which I found extremely wasteful, and double-pass of `initGlyphs` just to calculate line heights seemed overkill. So. I got rid of that.
* splitText got a huge facelift. 
  * Now allows setting custom `font` (required for proper word-wrap in HtmlText)
  * Can take optional `sizes` array in which it will store the line widths. Since we already calculate that stuff for word-wrapping, there is no reason to just throw it away if we'll need it down the line anyway.
  * Can take optional `prevChar` argument to ensure proper kerning when splitting multiple consequetive strings (for HtmlText).
  * If `sizes` were passed, it still will go trough the word-wrap cycle to calculate line widths.
* `Text.initGlyphs` signature got simplified to just `text:UString, rebuild:Bool` and now doesn't call itself.
  * There is a small performance loss when `align = Left` and `maxWidth` not set, becuase `splitText` runs it's routine, however due to just 2 (instead of 4) passes over the input it's still more optimal.
  * Got rid of `lines.shift()` in favor of index caret.
* `HtmlText`
  * introduced `imageCache`, which is used to store results of `loadImage` call (filled in `buildSizes` and retrieved in `addNode`)
  * `buildSizes` now actually does what the name implies - it builds the line sizes (*see below) array.
    * Now applies word-wrapping and caches the result.
  * Got rid of that `remainingSize` thing, because it's purpose was pretty cryptic and operated weirdly.
  * Implemented `<p align=''>` support. It doesn't work as HTML paragraph tag however and I'm not sure if it's worth implementing accurate behavior, since it complicates things a lot. Accurate behavior causes it to add empty line both before and after paragraph. Those lines are omitted when it's first/last `p` tag in the input, and chained `p` tags (`<p><p>text</p></p>`) would cause only one empty line before and after. I'm not doing that. :)

Regarding `buildSizes`. Currently there is a TODO and commented out heights argument. I decided to not implement it at the moment, but it is required in order to properly render multiple different fonts that have different heights. When implemented, buildSizes will also calculate line heights for that purpose.

**Comment by ncannasse** [comment]
These are always harder to write than to think about :)
Sad thing, there are harder to review as well. We have an important deadline soon so I'll this rest for a few weeks before merging. Don't hesitate to update with issues you might find in the meanwhile.

**Comment by Yanrishatum** [comment]
HtmlText: Added calculations of line heights for each individual line as well as it's baseline. Due to that, differently-sized fonts now won't overlap with lines above/below (+  tag).

**Comment by ncannasse** [comment]
@trethaller @D0pu can you test with DB+NG and merge if ok?

**Comment by D0pu** [comment]
There is some issue with the sample "htmlText" :

On the left side, there is what is expected. On the right, it's the result with this branch.
image (1)

In game, I found some problems when lines have image:
- Words before image are below the correct alignment, the image and the words after it are correctly placed.
- When the image is at the end of the line, the line is broken in the wrong place

**Comment by Yanrishatum** [comment]
@D0pu 
I've managed to reproduce some issues on htmlText, but not ones on the left part: 
image
I'll merge master into the branch and check again.

As for `` tag: Could you provide reproduction html? (Preferably with image/font files, but just saying their dimensions will do as well)

**Comment by Yanrishatum** [comment]
* Fixed `xMin` being incorrectly calculated. (HtmlText sample middle part)
* Fixed `calcHeight` and `calcSizeHeight` to use proper line height based on calculations instead of default font.
* Changed  height calculations to adjust baseline. This change relies on assumption that `baseLines` **always** lower than `lineHeight` in the fonts.
* Fixed vertical offset of text that word-wrapped (If first line was had different baseline to next ones)
* Added documentation to `load` methods.
* Added `defaultLoad` methods. This allows for less hassle when using HtmlText extensively and load methods use same code everywhere.
* Extended HtmlText sample: Added showcase of all supported tags.
* Changed how `<font face` behaved: If `loadFont` returns `null`, it uses currently active font and does not create extra `glyphs` instance.

Please retest. :)

**Comment by D0pu** [comment]
The sample works for me :)

I still have some issue :
```
var flow = createFlow(s2d);
flow.y = yoffset;
flow.x = 150;
flow.horizontalAlign = FlowAlign.Middle;
flow.layout = Vertical;
{
	var f1 = createFlow(flow);
	createText(f1, 'This is a very long line <font color="#ffff00">CreateBug</font> here', Align.Left);
	var f2 = createFlow(flow);
	createText(f2, 'This is a very long line CreateBug NotHere', Align.Left);
}
```
When a <font> has to go to the next line, the content after </font> goes to the new line too.


About images, for example in Northgard, the text is not align with the dot now (when there is an _img_ in the HtmlText) :

You can see a difference in the sample (maybe not the exact same context) :
```
var flow = createFlow(s2d);
flow.y = yoffset;
flow.x = 150;
flow.horizontalAlign = FlowAlign.Middle;
flow.layout = Vertical;
{
	var f1 = createFlow(flow);
	var text = createText(f1, 'This is a very long line <font color="#ffff00"><img src="logo"/>Center</font> here', Align.Left);
	text.loadImage = function( src : String ) : h2d.Tile {
		if ( src == "logo" ) {
			var t = hxd.Res.hxlogo.toTile();
			t.scaleToSize(16, 16);
			return t;
		}
		return null;
	}
}
```

**Comment by Yanrishatum** [comment]
Found the reason of incorrect line-break, fixed.

> About images, for example in Northgard, the text is not align with the dot now (when there is an img in the HtmlText)

This isn't particularly a bug, but a consequence of proper line-height calculation. Before, it was a constant based on an `HtmlText.font` values, while now it bases off: 
* Actual used fonts on the line (both height and baseline calculated based on biggest font).
* In case of `` - it puts the bottom of the image on the baseline, extending it if needed.

This produces more expected result from resulting render. However there is be an issue with `dx`/`dy`, as this calculation does not account for it. 
If preserving legacy behavior is important, probably worth adding flag that would exclude images from affecting height of the line.

**Comment by Yanrishatum** [comment]
* Fixed invalid line-breaks
* `dy` now accounted when calculating line heights with ``.
* Added `heaps_legacy_html_img` compilation flag. If set, images won't adjust baseLine/height as it's currently in master.

**Comment by trethaller** [comment]
Thanks! One minor thing, I'm pretty sure @ncannasse will prefer something like `var constantLineHeight = false;` in HtmlText, over a compilation switch for something so specific.

**Comment by Yanrishatum** [comment]
* Added `lineHeightMode` - 3 states, accurate (default), text only (images don't affect line heights) and constant (legacy)
* Removed compilation flag.
* Fixed line break height calculation (`<br/><br/>` produced just one line break instead of empty line)

I made it an enum mode, because `constantLineHeight` is 100% legacy behavior, and some people would want HtmlText to adjust to font heights, while not touching images.

**Comment by D0pu** [comment]
HTMLText looks good now.

I found one strange behaviour with `Text` and `\n`

Please, can you add and check the following test in the sample `Text` ?
```
yoffset += flow.getBounds().height + 35;

{
	var flow = createFlow(s2d);
	flow.y = yoffset;
	flow.horizontalAlign = FlowAlign.Left;
	flow.horizontalSpacing = 15;
	flow.layout = Horizontal;
	createText(flow, "LEFT: It is a text with a new line added after THAT\nto test the alignment", Align.Left);
	createText(flow, "CENTER: It is a text with a new line added after THAT\nto test the alignment", Align.Center);
	createText(flow, "RIGHT: It is a text with a new line added after THAT\nto test the alignment", Align.Right);
	createText(flow, "MULTICENTER: It is a text with a new line added after THAT\nto test the alignment", Align.MultilineCenter);
	createText(flow, "MULTIRIGHT: It is a text with a new line added after THAT\nto test the alignment", Align.MultilineRight);
}
```

**Comment by Yanrishatum** [comment]
Fixed.

**Comment by D0pu** [comment]
Nice !

Another thing, sorry...
```
yoffset += flow.getBounds().height + 10;

{
	var flow = createFlow(s2d);
	flow.debug = false;
	flow.y = yoffset;
	flow.horizontalAlign = FlowAlign.Left;
	flow.maxWidth = 766;
	flow.horizontalSpacing = 8;
	flow.layout = Horizontal;
	var f = createText(flow, "The image and the text are on <font color='#FF00FF'><img src='logo'/>the_same_line</font> bla bla bla.", Align.Left);
	f.maxWidth = 250;
	var f = createText(flow, "The comma is on the same line as <font color='#FF00FF'><img src='logo'/>that</font>, bla bla bla.", Align.Left);
	f.maxWidth = 235;
	yoffset += flow.getBounds().height + 10;
}
```

On top, it's your branch
image

For the first `HtmlText`, when the content of _\<font\>_ has no space, it should be on the same line.
For the second, the comma has to be on the same line (without space before the comma).

**Comment by Yanrishatum** [comment]
No worries, it's better to iron out all quirks before merge rather than hotpatch them later when sudden issues arise. :)
I'll look into it tomorrow, but comma issue will be pretty hard to solve, This is due to word-wrapping being handled separately for each text node, and comma could not fit into width restriction, and since it was not part of word "that" from technical standpoint, it got wrapped. `splitText` should somehow retrospectively know about attached word. 
And for first sample, I'm not really sure why it should be on next line. `` is not a text node, and line-breaks separately, if algorithm decided it can fit into end of the line - it fits it here.

**Comment by Yanrishatum** [comment]
Okay. Regarding text on the left: I'm not sure why that differs, apart from rewritten word-wrapping conditions which should've remained roughly the same and as accurate. 
And for right part: I can't really solve it without rewriting whole text processing again. `<font>text</font>,` causes it to be processed as separate `text` and `,`, which would cause wrapping even on old HtmlText, and difference with my solution is mainly because it does regard space just before `<font>` tag and not ignore it.

**Comment by ncannasse** [comment]
In current HtmlText, line breaks can only be added at places there is a space (according to font.isBreakChar). And it is important that either several img or an img followed by some text without spaces in between are maintained as a single unbreakable block.

I agree this differ from HTML but I would like to be able to keep this important property.

**Comment by Yanrishatum** [comment]
Welp, back to the drawing board then. :)

**Comment by ncannasse** [comment]
Tell me when it's ready for final review.

**Comment by Yanrishatum** [comment]
@ncannasse Yep. Fixed found bugs and fixed #725. In general I can say it's pretty much ready, as there's nothing else to add. There is one bug, however for an edge case of Text having EXTREMELY small maxWidth, in which case HtmlText would do the splitting where it shouldn't, but given the very big edge case - see image.
I'm not sure if I reimplemented merge correctly, however. As mentioned in comment - refactoring is required (And I already have an issue open about that ;) )

image
(maxWidth is smaller than "that" word, causing comma to be placed at a new line)

**Comment by ncannasse** [comment]
Seems correct since there is no space between the img and "that" so there
should not be split by newline

Le dim. 1 déc. 2019 à 13:44, Pavel Alexandrov <notifications@github.com> a
écrit :

**Comment by ncannasse** [comment]
@D0pu to test against DB/NG for regressions ;)

**Comment by D0pu** [comment]
After merge, I got an error with **<b>** with loadFont("bold")
```
public dynamic function loadFont( name : String ) : Font {
     return defaultLoadFont(name);	// replace with return font; ?
}
```

After this fix, everything looks great

**Comment by Yanrishatum** [comment]
Fixed. Now it would fallback to current Font instance when `defaultLoadFont` yields `null`.

File: h2d/HtmlText.hx [hunk 1]
```diff
@@ -137,39 +250,130 @@ class HtmlText extends Text {
...
+					makeLineBreak();
+				}
 			default:
 			}
 		} else {
 			newLine = false;
 			var text = htmlToText(e.nodeValue);
-			var xp = 0.;
-			for( i in 0...text.length ) {
+			var leftMargin = lines.length == 0 ? 0 : lines.pop();
```
Comment (trethaller) [inline h2d/HtmlText.hx]: for example here I find the `pop` on lines and not other arrays a bit worrying, but I trust that the logic is right ^^

File: h2d/HtmlText.hx [hunk 2]
```diff
@@ -39,91 +75,168 @@ class HtmlText extends Text {
...
+			}
+			splitNode.node = null;
+			return w;
+		}
+
 		if( e.nodeType == Xml.Element ) {
-			var len = 0.;
+
+			inline function makeLineBreak() {
+				var fontInfo = lineHeightMode == Constant ? this.font : font;
```
[the remark this answers is not in the document]
Comment (Yanrishatum) [reply h2d/HtmlText.hx]: `Constant` uses default font when making line breaks as to ensure constant line height. While all other modes adapt line height so lines won't overlap if fonts of different height are used.

File: h2d/HtmlText.hx (hunk 1 above)
Comment (Yanrishatum) [reply h2d/HtmlText.hx]: It's removed from the array because it gets modified during processing and inserted back (or in case if word-wrap occurs). There is no need to pop other metrics, as line remains here, only width is changed, and if line-break happens, those metrics are populated with appropriate values.

File: h2d/HtmlText.hx (hunk 2 above)
[the remark this answers is not in the document]
Comment (trethaller) [reply h2d/HtmlText.hx]: gotcha, seems obvious now

File: h2d/HtmlText.hx [hunk 3]
```diff
@@ -39,91 +75,168 @@ class HtmlText extends Text {
...
 		calcWidth = xMax - xMin;
-		calcHeight = y + font.lineHeight;
-		calcSizeHeight = y + (font.baseLine > 0 ? font.baseLine : font.lineHeight);
+		calcHeight = y + heights[sizePos];
+		calcSizeHeight = y + baseLines[sizePos];//(font.baseLine > 0 ? font.baseLine : font.lineHeight);
 		calcDone = true;
 	}
 
-	function buildSizes( e : Xml, font : Font, sizes : Array<Float>, forSplit ) {
+	function buildSizes( e : Xml, font : Font, lines : Array<Float>, heights : Array<Float>, baseLines : Array<Float>, splitNode:SplitNode ) {
```
[the remark this answers is not in the document]
Comment (Yanrishatum) [reply h2d/HtmlText.hx]: Will be unified into singular structure and renamed as `metrics`, which is pretty descriptive of what it contains.

**Comment by Yanrishatum** [comment]
While doing the unification from 3 array to singular structure array I've found bug with word-wrapping when it comes to break-chars (e.g. Asian alphabets).
* `Charset.isCJK` check does not cover all CJK, so only some of them are considered as break chars. Below sample assume CJK detection is correct.
* In case of `東風ゴシ<span>ック</span>` algorithm would not trigger line-break between `シッ` characters while it should be able to do so.

**Comment by ncannasse** [comment]
@Yanrishatum is the bug a consequence of the patch ? if not, it should be fixed separately so in case of regression we can test against each patch independently

**Comment by Yanrishatum** [comment]
@ncannasse CJK detection is not relevant to this PR, but as I noted in DMs with Trethaller, it can be mitigated with expanding CJK detection following unicode specification: https://www.unicode.org/reports/tr14/tr14-34.html#ID
As for second issue: It's still present on Master, but in PR behavior differs and can be considered a regression. I will leave out CJK detection fix for later PR, but I still need to fix break char behavior.
Gifs I recorder for Trethaller as reference:
Master:
master
This PR:
pr
How it roughly expected to work (with CJK and line-breaks fixed, what shown is not a proper fix):
expected

**Comment by ncannasse** [comment]
Agreed

Le jeu. 9 janv. 2020 à 02:39, Pavel Alexandrov <notifications@github.com> a
écrit :

**Comment by Yanrishatum** [comment]
Should be fixed and ready now. 
Tested next edge cases to ensure issue is resolved:
```haxe

			// Test: first word too long split: Have previous splittable word in other node
			'Split before TH<font>IS word</font>', // PASS
			
			// Test: first word too long split: No previous splittable word in other node
			'Split_after_TH<font>IS word</font>', // PASS
			
			// Test: first word too long split: No break char 'till text end
			'Split before TH<font>IS_word</font>', // PASS

			// [CJK] Just CJK
			"あいうえおかき東風ゴシック", // FAIL: Incorrect CJK detection

			// [CJK] Just CJK (guaranteed split)
			"東風ゴ", // PASS
			
			// [CJK] CJK split between spans
			"東風<span>ゴ</span>", // PASS
```

**Comment by ncannasse** [comment]
Merged, thanks ! I will made some additional change, which is to remove public from splitRawText as I don't see it usable as public API (we already have splitText for this)
